cmd/state-svc/internal/deprecation/deprecation.go

```diff
@@ -0,0 +1,187 @@
+package deprecation
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ActiveState/cli/internal/constants"
+	constvers "github.com/ActiveState/cli/internal/constants/version"
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/graph"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/multilog"
+	"github.com/hashicorp/go-version"
+	"github.com/patrickmn/go-cache"
+)
+
+const (
+	// DefaultTimeout defines how long we should wait for a response from constants.DeprecationInfoURL
+	DefaultTimeout = time.Second
+
+	cacheKey = "info"
+)
+
+type ErrTimeout struct {
+	error
+}
+
+// info details deprecation information for a given version
+type info struct {
+	Version     string `json:"version"`
+	versionInfo *version.Version
+	Date        time.Time `json:"date"`
+	DateReached bool
+	Reason      string `json:"reason"`
+}
+
+// Checker is the struct that we use to do checks with
+type Checker struct {
+	timeout time.Duration
+	config  configurable
+	cache   *cache.Cache
+}
+
+// configurable defines the configuration function used by the functions in this package
+type configurable interface {
+	ConfigPath() string
+	GetTime(key string) time.Time
+	Set(key string, value interface{}) error
+	Close() error
+}
+
+// NewChecker returns a new instance of the Checker struct
+func NewChecker(timeout time.Duration, configuration configurable) *Checker {
+	return &Checker{
+		timeout,
+		configuration,
+		cache.New(15*time.Minute, 5*time.Minute),
+	}
+}
+
+// Check will Check if the current version of the tool is deprecated and returns deprecation info if it is.
+// This uses a fairly short timeout to Check against our deprecation url, so this should not be considered conclusive.
+func (checker *Checker) Check() (*graph.DeprecationInfo, error) {
+	if !constvers.NumberIsProduction(constants.Version) {
+		return nil, nil
+	}
+
+	versionInfo, err := version.NewVersion(constants.Version)
+	if err != nil {
+		return nil, errs.Wrap(err, "Invalid version number: %s", constants.Version)
+	}
+
+	data, exists := checker.cache.Get(cacheKey)
+	if !exists {
+		go func() {
+			err := checker.updateDeprecationInfo()
+			if err != nil {
+				multilog.Critical("Could not update deprecation information %s", errs.JoinMessage(err))
+			}
+		}()
+
+		return nil, nil
+	}
```

Comment:
This is basically ensuring that some calls won't get deprecation info simply because we haven't warmed the cache. I think we should warm the cache in the background, on a timer. If there are no network issues clients should get deprecation info 100% of the time.